### PR TITLE
Add API to iterate over items in a JSON collection

### PR DESCRIPTION
**Pull request by dan4thewin**
This PR includes unit tests with 100% branch coverage, a CBMC proof with 100% coverage and no errors, and doxygen updates.

### CODE REVIEW

File: source/core_json.c [hunk 1]
```diff
@@ -1676,3 +1676,131 @@ JSONStatus_t JSON_SearchT( char * buf,
...
+            default:
+                ret = JSONIllegalDocument;
+                break;
+        }
+    }
+
+    if( found == true )
+    {
+        ret = JSONSuccess;
+        ( void ) skipSpaceAndComma( buf, next, max );
```
Comment (yourslab) [inline source/core_json.c:1744]: Should the return value of this function be ignored? I'm guessing that this function should not handle any validation and `JSON_Validate` will be responsible instead, so just making sure.

File: source/include/core_json.h [hunk 2]
```diff
@@ -231,4 +232,93 @@ JSONStatus_t JSON_SearchConst( const char * buf,
...
+ * and #JSONSuccess is returned; otherwise the structure is unchanged and #JSONNotFound
+ * is returned.
+ *
+ * @param[in] buf  The buffer to search.
+ * @param[in] max  size of the buffer.
+ * @param[in,out] start  The index at which the collection begins.
+ * @param[in,out] next  The index at which to seek the next value.
+ * @param[out] outPair  A pointer to receive the next key-value pair.
+ *
+ * @note This function expects a valid JSON document; run JSON_Validate() first.
```
Comment (yourslab) [inline source/include/core_json.h:265]: Whoops I missed this here

File: test/unit-test/core_json_utest.c [hunk 3]
```diff
@@ -935,6 +940,100 @@ void test_JSON_Search_Legal_Array_Documents( void )
...
+/**
+ * @brief Test that JSON_Iterate returns the given values from a JSON array.
+ */
+void test_JSON_Iterate_Legal_Array_Documents( void )
+{
+    JSONStatus_t jsonStatus;
+    size_t start = 0, next = 0;
+    JSONPair_t pair = { 0 };
+
+#define doIterate( type, answer )                                    \
```
Comment (yourslab) [inline test/unit-test/core_json_utest.c:952]: Just a nit about the name
```suggestion
#define iterateObject( type, answer )                                    \
```

File: test/unit-test/core_json_utest.c [hunk 4]
```diff
@@ -935,6 +940,100 @@ void test_JSON_Search_Legal_Array_Documents( void )
...
+/**
+ * @brief Test that JSON_Iterate returns the given keys and values from a JSON object.
+ */
+void test_JSON_Iterate_Legal_Object_Documents( void )
+{
+    JSONStatus_t jsonStatus;
+    size_t start = 0, next = 0;
+    JSONPair_t pair = { 0 };
+
+#define doIterate2( key_, type, answer )                             \
```
Comment (yourslab) [inline test/unit-test/core_json_utest.c:991]:
```suggestion
#define iterateObject( key_, type, answer )                             \
```

**Review by yourslab — APPROVED** [review]

**Review by yourslab — DISMISSED** [review]
Just a few nits. This looks good to me already.

**Review by yourslab — APPROVED** [review]

File: source/core_json.c (hunk 1 above)
Comment (dan4thewin) [reply source/core_json.c:1744]: just so

**Review by sarenameas — APPROVED** [review]

**Review by cobusve — APPROVED** [review]

**Pull request merged**